client/app/components/tags-control/TagsControl.jsx

```diff
@@ -18,87 +21,118 @@ export default class TagsControl extends React.Component {
+  };
 
   closeEditModal = () => {
     this.setState({ showModal: false });
-  }
-
-  // eslint-disable-next-line class-methods-use-this
-  renderPrependTags() {
-    return null;
-  }
-
-  renderTags() {
-    return map(this.props.tags, tag => (
-      <span className="label label-tag" key={tag} title={tag}>{tag}</span>
-    ));
-  }
+  };
 
-  // eslint-disable-next-line class-methods-use-this
-  renderAppendTags() {
-    return null;
-  }
+  onTagsChanged = (newTags) => {
+    this.props.onEdit(newTags);
+    this.closeEditModal();
+  };
 
   renderEditButton() {
-    if (!this.props.canEdit) {
-      return null;
-    }
-
     const tags = map(this.props.tags, trim);
-
     const buttonLabel = tags.length > 0
       ? <i className="zmdi zmdi-edit" />
-      : <Fragment><i className="zmdi zmdi-plus" /> Add tag</Fragment>;
+      : <React.Fragment><i className="zmdi zmdi-plus m-r-5" />Add tag</React.Fragment>;
 
     return (
-      <Fragment>
-        <a className="label label-tag" role="none" onClick={this.openEditModal}>
-          {buttonLabel}
-        </a>
+      <React.Fragment>
+        <a className="label label-tag" role="none" onClick={this.openEditModal}>{buttonLabel}</a>
         {this.state.showModal && (
           <TagsEditorModal
             tags={tags}
-            availableTags={this.availableTags}
-            close={this.onTagsChanged}
-            dismiss={this.closeEditModal}
+            availableTags={this.state.availableTags}
+            onConfirm={this.onTagsChanged}
+            onCancel={this.closeEditModal}
           />
         )}
-      </Fragment>
+      </React.Fragment>
     );
   }
 
   render() {
+    const children = filter(React.Children.toArray(this.props.children), isObject);
     return (
       <div className={'tags-control ' + this.props.className}>
-        {this.renderPrependTags()}
-        {this.renderTags()}
-        {this.renderAppendTags()}
-        {this.renderEditButton()}
+        {filter(children, child => child.type === this.constructor.Prepend)}
+        {map(this.props.tags, tag => (
+          <span className="label label-tag" key={tag} title={tag}>{tag}</span>
+        ))}
+        {filter(children, child => child.type === this.constructor.Append)}
+        {this.props.canEdit && this.renderEditButton()}
       </div>
     );
   }
 }
+
+function modelTagsControl({ archivedTooltip }) {
+  // See comment for `propTypes`/`defaultProps`
+  // eslint-disable-next-line react/prop-types
+  function ModelTagsControl({ isDraft, isArchived, ...props }) {
+    return (
+      <TagsControl {...props}>
+        <TagsControl.Prepend>
+          {!isArchived && isDraft && (
+            <span className="label label-tag-unpublished">Unpublished</span>
+          )}
+          {isArchived && (
+            <Tooltip placement="right" title={archivedTooltip}>
+              <span className="label label-tag-archived">Archived</span>
+            </Tooltip>
+          )}
+        </TagsControl.Prepend>
```

Comment:
Instead of all the append/prepend logic, why not render the two special tags and then call `<TagsControl>`?

```jsx
render() {
  return (
    <React.Fragment>
        {...draft label...}
        {...archive label...}
        <TagsControl ...props/>
    </React.Fragment>
  );
}
```

Also, how about turning ModelTagsControl into a regular component that accepts an `archiveToolTip` prop and then QueryTagsControl/DashboardsTagsControl become:

```jsx
function QueryTagsControl(props) {
    return <ModalTagsControl archiveTooltip="..." ...props/>
}
```

WDYT?

Reply:
> Instead of all the append/prepend logic, why not render the two special tags and then call `<TagsControl>`?

Because it will break nesting level of elements (new elements will be outside of wrapper).

> Also, how about turning ModelTagsControl into a regular component that accepts an `archiveToolTip` prop

Indeed it will work as expected, but there will be no huge difference with current implementation; HOC just hides that intermediate component.

Reply:
@ranbena What's you opinion about this?

Reply:
> Because it will break nesting level of elements (new elements will be outside of wrapper).

Right. 👍 It seems we don't use `TagsControl.Append` anywhere... Maybe use `children` directly to avoid the extra complexity?

> Indeed it will work as expected, but there will be no huge difference with current implementation; HOC just hides that intermediate component.

The main difference is a more familiar pattern.

Reply:
I favor @arikfr's suggestion cause it's more readable
```jsx
function QueryTagsControl(props) {
  return <ModalTagsControl archiveTooltip="..." ...props/>
}
```

Regarding prepend - is append used at all?
Doesn't look like it.. so why not just go with:
```jsx
<TagsControl ...props/>
  {...draft label...}
  {...archive label...}
</TagsControl>
```
And then:
```jsx
render() {
  return (
    <div className={'tags-control ' + this.props.className}>
      {this.props.children}
      {this.props.tags.map...}
      {this.renderEditButton()}
    </div>
  );
}
```

Reply:
When I wrote this ^^ I didn't realize arik already suggested it...